collect_app/src/main/java/org/odk/collect/android/jobs/AutoSendWorker.java

```diff
@@ -1,81 +1,78 @@
 import java.util.HashMap;
 import java.util.Iterator;
 import java.util.Map;
 import java.util.Set;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+
+import androidx.work.Worker;
+import timber.log.Timber;
 
 import static org.odk.collect.android.provider.FormsProviderAPI.FormsColumns.AUTO_SEND;
 import static org.odk.collect.android.utilities.ApplicationConstants.RequestCodes.FORMS_UPLOADED_NOTIFICATION;
 
-public class NetworkReceiver extends BroadcastReceiver implements InstanceUploaderListener {
-
-    // turning on wifi often gets two CONNECTED events. we only want to run one thread at a time
-    public static boolean running;
-    InstanceServerUploader instanceServerUploader;
+public class AutoSendWorker extends Worker implements InstanceUploaderListener {
+    public static final String TAG = "autoSendWorker";
 
-    InstanceGoogleSheetsUploader instanceGoogleSheetsUploader;
+    private InstanceServerUploader instanceServerUploader;
+    private InstanceGoogleSheetsUploader instanceGoogleSheetsUploader;
 
     private String resultMessage;
+    private CountDownLatch countDownLatch;
+    Result workResult;
 
+    @NonNull
     @Override
-    public void onReceive(Context context, Intent intent) {
-        // make sure sd card is ready, if not don't try to send
+    public Result doWork() {        // make sure sd card is ready, if not don't try to send
         if (!Environment.getExternalStorageState().equals(Environment.MEDIA_MOUNTED)) {
-            return;
+            return Result.FAILURE;
         }
 
-        String action = intent.getAction();
-        ConnectivityManager manager = (ConnectivityManager) context.getSystemService(
+        countDownLatch = new CountDownLatch(1);
+
+        ConnectivityManager manager = (ConnectivityManager) getApplicationContext().getSystemService(
                 Context.CONNECTIVITY_SERVICE);
         NetworkInfo currentNetworkInfo = manager.getActiveNetworkInfo();
 
-        if (action.equals(ConnectivityManager.CONNECTIVITY_ACTION)) {
-            if (currentNetworkInfo != null
-                    && currentNetworkInfo.getState() == NetworkInfo.State.CONNECTED) {
-                uploadForms(context, isFormAutoSendOptionEnabled(currentNetworkInfo));
-            }
-
-            ServerPollingJob.pollServerIfNeeded();
-        } else if (action.equals("org.odk.collect.android.FormSaved")) {
-            ConnectivityManager connectivityManager = (ConnectivityManager) context
-                    .getSystemService(Context.CONNECTIVITY_SERVICE);
-            NetworkInfo ni = connectivityManager.getActiveNetworkInfo();
+        uploadForms(getApplicationContext(), isFormAutoSendOptionEnabled(currentNetworkInfo));
 
-            if (ni != null && ni.isConnected()) {
-                uploadForms(context, isFormAutoSendOptionEnabled(currentNetworkInfo));
-            }
+        try {
+            countDownLatch.await(2, TimeUnit.MINUTES);
```

Comment:
So I just put a random value here, where I am saying if the app doesn't get a result from the `InstanceUploader` after two minutes then it should simply just return a failure. What I want help with is selecting a realistic time that should be used based on how long forms have taken to upload in the past due to size and connectivity and/or if the `InstanceUploaer` has any time out mechanism itself that would ensure this condition is never met.

Reply:
I could always use the other version of the `await` method that doesn't take a time. I just wanted to put a time as a fail safe so the job can fail if necessary and hand over the result to the `WorkManager`.

Reply:
`WorkManager` can run a job immediately in addition to based on a trigger so I think the eventual goal would be to replace the existing `AsyncTask`s at some point. Does that sound right to you, @jd-alexander? Having the actual upload happen here would mean not having to coordinate with yet another thread so no countdown latch or anything like this. 

`AsyncTask`s are only supposed to be for short-running tasks, right? Now that I think more about it, I'm surprised that uploads were implemented this way and that it doesn't seem to cause problems. I suppose one of the big problems with `AsyncTask`s are the coordination with the `Activity` and since the manual upload UI is blocking and autosend isn't triggered by an activity, perhaps things running longer is ok. Anyway, I assume you decided to keep the `AsyncTask`s for now to reduce the risk of regressions and that's probably a good way to start.

What would it take to know for sure whether there are any parts of the upload process that can hang indefinitely? What would be the implications of that happening? It would be good to know whether a `Worker` eventually gets killed by the system. If there are parts of the upload process that can hang indefinitely, this is a problem with the current implementation as well, right?

What are the implications of killing the job after 2 minutes? Let's say there are lots of forms with lots of media that need to be sent. The job would fail. Then it would get re-enqueued when another form were finalized or the connectivity status changed. This could be a long time in the future so if we do decide to have a timeout I think we should send a notification to the user and let them know they can manually trigger a send (if the send screen isn't hidden in admin settings).

My current leaning would be to not have the timeout. It seems that most closely matches the current behavior and I don't think there are any terrible implications though it would be good to verify that and to at least try to verify if/where there's an opportunity for something to hang.

Reply:
>   Anyway, I assume you decided to keep the AsyncTasks for now to reduce the risk of regressions and that's probably a good way to start.

Yes, that's what I was thinking. Also, the `Workers` dont have proper support for asynchronous work so that's a question I am going to pose as well to get some answers from those who are shaping the current `WorkManager` API. 

> What would it take to know for sure whether there are any parts of the upload process that can hang indefinitely?

I know upload tend to hang if extremely large files are being uploaded but normally the `HttpClient` being used takes care of that due to timeouts and size limitations. I could check with @xsteelej and just ask him if he has any ideas around this since we was in the `Http` code for a long while. 

> What would be the implications of that happening? It would be good to know whether a Worker eventually gets killed by the system. 

I posted an issue for the `PeriodicWorkRequest` behaviour that I have been trying to verify so if I get a response from the Android arch team  I can pose these questions.

> If there are parts of the upload process that can hang indefinitely, this is a problem with the current implementation as well, right?

Yes because the worker is simply starting the `ServerUploader` that's wrapped in the  `AsyncTask`.